File: cognipy/CogniPyLib/Taxonomy/TaxonomyNode.cs
using CogniPy.Collections;
using System;
using System.Collections.Generic;
using System.ComponentModel;

namespace CNL.NET
{
    public delegate void RefreshNeededEventHandler(object sender, EventArgs e);

    public abstract class TaxonomyNode : INotifyPropertyChanged, IComparable
    {
        public event PropertyChangedEventHandler PropertyChanged;
        IInvokableProvider invokableProvider;
        public int CompareTo(object obj)
        {
            if (obj is TaxonomyNode)
                return ENText.CompareTo((obj as TaxonomyNode).ENText);
            else
                return 0;
        }

        protected TaxonomyNode(IInvokableProvider invokableProvider)
        {
            this.invokableProvider = invokableProvider;
            this.children = new SortedObservableCollection<TaxonomyNode>(invokableProvider);

        }

        Guid _uniqueId = Guid.NewGuid();
        public Guid ID { get { return _uniqueId; } }

        TaxonomyNode _parentNode = null;
        public TaxonomyNode Parent { get { return _parentNode; } set { _parentNode = value; } }
        bool _isOntorionMode = false;
        public bool IsInOntorionModule
        {
            get { return _isOntorionMode; }
            set
            {
                _isOntorionMode = value;
                notifyPropertyChanged("IsInOntorionModule");

            }
        }
        bool _isExpanded = false;
        public bool isExpandedNoPropertyChanged
        {
            set { _isExpanded = value; }
        }

        public bool IsExpanded
        {
            get { return _isExpanded; }
            set
            {
                _isExpanded = value;
                notifyPropertyChanged("IsExpanded");

            }
        }

        public void SetupChild(TaxonomyNode node)
        {
            if (node.ENText != "" || node.MyType == NodeType.FictiousChild)
            {
                node.SetInvokableProvider(invokableProvider);
                Children.Add(node);
            }
        }

        void SetInvokableProvider(IInvokableProvider invokableProvider2)
        {
            this.invokableProvider = invokableProvider2;
            Children.SetInvokableProvider(invokableProvider2);
            foreach (var c in Children)
                c.SetInvokableProvider(invokableProvider2);
        }

        public void RemoveChild(TaxonomyNode node)
        {
            if (this.Children.Contains(node))
            {
                Children.Remove(node);
            }
        }

        public HashSet<string> names = new HashSet<string>();

        public void SetupName(string name)
        {
            if (names.Add(name))
            {
                purgeCache();
                notifyPropertyChanged("ENText");

            }
        }

        public void RemoveName(string name)
        {
            if (names.Remove(name))
            {
                purgeCache();
                notifyPropertyChanged("ENText");

            }
        }

        public bool ContainsName(string name)
        {
            return names.Contains(name);
        }

        public void CopyNames(TaxonomyNode node)
        {
            bool r = false;
            foreach (var n in node.names)
            {
                r |= names.Add(n);
            }
            if (r)
            {
                purgeCache();
                notifyPropertyChanged("ENText");
            }
        }

        public CogniPy.ARS.EntityKind Kind { get; set; }

        void purgeCache()
        {
            cachedENText = null;
        }
        protected string cachedENText = null;
        public abstract string ENText { get; }


        public enum NodeType { Standard, TopConcept, TopRole, TopAttribute, Nothing, FictiousChild, OntorionSynced, OntorionNotSynced }

        string _prefix = "";
        public string Prefix
        {
            protected set { _prefix = value; }
            get { return _prefix; }
        }

        public bool IsImported { get; set; }
        NodeType _myType = NodeType.Standard;
        public NodeType MyType
        {
            get { return _myType; }
            set
            {
                _myType = value;
                notifyPropertyChanged("MyType");
            }
        }

        public Object Self { get { return this; } }

        private SortedObservableCollection<TaxonomyNode> children;
        public SortedObservableCollection<TaxonomyNode> Children
        {
            get { return children; }
            set
            {
                children = value;
                notifyPropertyChanged("Children");
            }
        }

        public static bool haveCommonName(TaxonomyNode a, TaxonomyNode b)
        {
            foreach (var A in a.names)
            {
                if (!b.names.Contains(A))
                    return false;
            }
            foreach (var B in b.names)
            {
                if (!a.names.Contains(B))
                    return false;
            }
            return true;
        }

        bool _hasAnnotation = false;
        public bool HasAnnotation
        {
            get { return _hasAnnotation; }
            set
            {
                _hasAnnotation = value;
                notifyPropertyChanged("HasAnnotation");
            }
        }

        private void notifyPropertyChanged(string propName)
        {
            if (this.invokableProvider.GetInvokable() != null)
            {
                this.invokableProvider.GetInvokable().Invoke(() =>
                {

                    if (this.PropertyChanged != null)
                    {
                        this.PropertyChanged(this, new PropertyChangedEventArgs(propName));
                        this.PropertyChanged(this, new PropertyChangedEventArgs("Self"));
                    }
                });
            }
        }
    }
}

File: cognipy/cognipylib/CNL/net/TaxonomyNode.java
package cnl.net;

import cognipy.collections.*;
import java.util.*;

public abstract class TaxonomyNode implements INotifyPropertyChanged, java.lang.Comparable
{
	public tangible.Event<PropertyChangedEventHandler> PropertyChanged = new tangible.Event<PropertyChangedEventHandler>();
	private IInvokableProvider invokableProvider;
	public final int compareTo(Object obj)
	{
		if (obj instanceof TaxonomyNode)
		{
			return getENText().compareTo((obj instanceof TaxonomyNode ? (TaxonomyNode)obj : null).getENText());
		}
		else
		{
			return 0;
		}
	}

	protected TaxonomyNode(IInvokableProvider invokableProvider)
	{
		this.invokableProvider = invokableProvider;
		this.children = new SortedObservableCollection<TaxonomyNode>(invokableProvider);

	}

	private UUID _uniqueId = UUID.NewGuid();
	public final UUID getID()
	{
		return _uniqueId;
	}

	private TaxonomyNode _parentNode = null;
	public final TaxonomyNode getParent()
	{
		return _parentNode;
	}
	public final void setParent(TaxonomyNode value)
	{
		_parentNode = value;
	}
	private boolean _isOntorionMode = false;
	public final boolean getIsInOntorionModule()
	{
		return _isOntorionMode;
	}
	public final void setIsInOntorionModule(boolean value)
	{
		_isOntorionMode = value;
		notifyPropertyChanged("IsInOntorionModule");

	}
	private boolean _isExpanded = false;
	public final void setIsExpandedNoPropertyChanged(boolean value)
	{
		_isExpanded = value;
	}

	public final boolean getIsExpanded()
	{
		return _isExpanded;
	}
	public final void setIsExpanded(boolean value)
	{
		_isExpanded = value;
		notifyPropertyChanged("IsExpanded");

	}

	public final void SetupChild(TaxonomyNode node)
	{
		if (!node.getENText().equals("") || node.getMyType() == NodeType.FictiousChild)
		{
			node.SetInvokableProvider(invokableProvider);
			getChildren().add(node);
		}
	}

	private void SetInvokableProvider(IInvokableProvider invokableProvider2)
	{
		this.invokableProvider = invokableProvider2;
		getChildren().SetInvokableProvider(invokableProvider2);
		for (TaxonomyNode c : getChildren())
		{
			c.SetInvokableProvider(invokableProvider2);
		}
	}

	public final void RemoveChild(TaxonomyNode node)
	{
		if (this.getChildren().contains(node))
		{
			getChildren().remove(node);
		}
	}

	public HashSet<String> names = new HashSet<String>();

	public final void SetupName(String name)
	{
		if (names.add(name))
		{
			purgeCache();
			notifyPropertyChanged("ENText");

		}
	}

	public final void RemoveName(String name)
	{
		if (names.remove(name))
		{
			purgeCache();
			notifyPropertyChanged("ENText");

		}
	}

	public final boolean ContainsName(String name)
	{
		return names.contains(name);
	}

	public final void CopyNames(TaxonomyNode node)
	{
		boolean r = false;
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
		for (var n : node.names)
		{
			names.add(n);
			r |= names.size() - 1;
		}
		if (r)
		{
			purgeCache();
			notifyPropertyChanged("ENText");
		}
	}

	private cognipy.ars.EntityKind Kind = cognipy.ars.EntityKind.values()[0];
	public final cognipy.ars.EntityKind getKind()
	{
		return Kind;
	}
	public final void setKind(cognipy.ars.EntityKind value)
	{
		Kind = value;
	}

	private void purgeCache()
	{
		cachedENText = null;
	}
	protected String cachedENText = null;
	public abstract String getENText();


	public enum NodeType
	{
		Standard,
		TopConcept,
		TopRole,
		TopAttribute,
		Nothing,
		FictiousChild,
		OntorionSynced,
		OntorionNotSynced;

		public static final int SIZE = java.lang.Integer.SIZE;

		public int getValue()
		{
			return this.ordinal();
		}

		public static NodeType forValue(int value)
		{
			return values()[value];
		}
	}

	private String _prefix = "";
	protected final void setPrefix(String value)
	{
		_prefix = value;
	}
	public final String getPrefix()
	{
		return _prefix;
	}

	private boolean IsImported;
	public final boolean getIsImported()
	{
		return IsImported;
	}
	public final void setIsImported(boolean value)
	{
		IsImported = value;
	}
	private NodeType _myType = NodeType.Standard;
	public final NodeType getMyType()
	{
		return _myType;
	}
	public final void setMyType(NodeType value)
	{
		_myType = value;
		notifyPropertyChanged("MyType");
	}

	public final Object getSelf()
	{
		return this;
	}

	private SortedObservableCollection<TaxonomyNode> children;
	public final SortedObservableCollection<TaxonomyNode> getChildren()
	{
		return children;
	}
	public final void setChildren(SortedObservableCollection<TaxonomyNode> value)
	{
		children = value;
		notifyPropertyChanged("Children");
	}

	public static boolean haveCommonName(TaxonomyNode a, TaxonomyNode b)
	{
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
		for (var A : a.names)
		{
			if (!b.names.contains(A))
			{
				return false;
			}
		}
//C# TO JAVA CONVERTER TODO TASK: There is no equivalent to implicit typing in Java unless the Java 10 inferred typing option is selected:
		for (var B : b.names)
		{
			if (!a.names.contains(B))
			{
				return false;
			}
		}
		return true;
	}

	private boolean _hasAnnotation = false;
	public final boolean getHasAnnotation()
	{
		return _hasAnnotation;
	}
	public final void setHasAnnotation(boolean value)
	{
		_hasAnnotation = value;
		notifyPropertyChanged("HasAnnotation");
	}

	private void notifyPropertyChanged(String propName)
	{
		if (this.invokableProvider.GetInvokable() != null)
		{
			this.invokableProvider.GetInvokable().Invoke(() ->
			{

					if (this.PropertyChanged != null)
					{
						for (PropertyChangedEventHandler listener : PropertyChanged.listeners())
						{
							listener.invoke(this, new PropertyChangedEventArgs(propName));
						}
						for (PropertyChangedEventHandler listener : PropertyChanged.listeners())
						{
							listener.invoke(this, new PropertyChangedEventArgs("Self"));
						}
					}
			});
		}
	}
}